MODEL TRAINING

In [31]:
import pandas as pd
import numpy as np
import json
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, accuracy_score, roc_auc_score
from sklearn.model_selection import train_test_split
from xgboost import XGBClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, roc_auc_score
import joblib


Leemos el dataset limpio

In [32]:
df = pd.read_csv('../data/MLA_100k_clean.csv')

Revisamos las columnas que hay en el dataset para luego generar los modelos.

In [33]:
df.head()

condition  base_price   price buying_mode  last_updated  \
0          1        80.0    80.0  buy_it_now      20150905   
1          0      2650.0  2650.0  buy_it_now      20150926   
2          0        60.0    60.0  buy_it_now      20150909   
3          1       580.0   580.0  buy_it_now      20151005   
4          0        30.0    30.0  buy_it_now      20150828   

   accepts_mercadopago  date_created  stop_time  status  initial_quantity  \
0                 True      20150905   20151104  active                 1   
1                 True      20150926   20151125  active                 1   
2                 True      20150909   20151108  active                 1   
3                 True      20150928   20151204  active                 1   
4                 True      20150824   20151023  active                 1   

   start_time  sold_quantity  available_quantity  
0    20150905              0                   1  
1    20150926              0                   1  
2    20150909              0                   1  
3    20150928              0                   1  
4    20150824              0                   1

generar los modelos de predicción utilizando diferentes algoritmos: Regresión Logística, RandomForestClassifier, XGBClassifier, Gradient Boosting y K-Nearest Neighbors (KNN).

In [34]:
# Convertir variables categóricas a variables dummy (One-Hot Encoding)
df_encoded = pd.get_dummies(df, drop_first=True)

# Definir X (todas las columnas excepto la columna objetivo) e y (columna objetivo)
X = df_encoded.drop('condition', axis=1)  # Asumiendo que 'condition' es tu variable objetivo
y = df_encoded['condition']

# Dividir el dataset en entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


#### Entrenamos los Modelos

Regresión Logística


In [35]:

# Entrenar el modelo de Regresión Logística
log_reg = LogisticRegression(max_iter=1000)
log_reg.fit(X_train, y_train)

# Hacer predicciones
y_pred_log = log_reg.predict(X_test)

# Calcular la precisión
accuracy_log = accuracy_score(y_test, y_pred_log)
print(f'Accuracy Logistic Regression: {accuracy_log}')

Accuracy Logistic Regression: 0.7095


RandomForestClassifier

In [36]:
# Crear el modelo Random Forest
rf_clf = RandomForestClassifier(n_estimators=100, random_state=42)

# Entrenar el modelo
rf_clf.fit(X_train, y_train)

# Hacer predicciones
y_pred_rf = rf_clf.predict(X_test)

# Evaluar el modelo
accuracy_rf = accuracy_score(y_test, y_pred_rf)
print(f'Accuracy Random Forest: {accuracy_rf}')


Accuracy Random Forest: 0.76685


XGBClassifier

In [37]:
# Crear el modelo XGBoost
xgb_clf = XGBClassifier(eval_metric='mlogloss')

# Entrenar el modelo
xgb_clf.fit(X_train, y_train)

# Hacer predicciones
y_pred_xgb = xgb_clf.predict(X_test)

# Evaluar el modelo
accuracy_xgb = accuracy_score(y_test, y_pred_xgb)
print(f'Accuracy XGBoost: {accuracy_xgb}')


Accuracy XGBoost: 0.78515


Gradient Boosting

In [44]:
# Entrenar Gradient Boosting
gb_clf = GradientBoostingClassifier(random_state=42)
gb_clf.fit(X_train, y_train)
y_pred_gb = gb_clf.predict(X_test)

# Evaluar Gradient Boosting
print("Gradient Boosting - Accuracy:", accuracy_score(y_test, y_pred_gb))

Gradient Boosting - Accuracy: 0.76995


K-Nearest Neighbors (KNN)

In [46]:
# Entrenar KNN
knn_clf = KNeighborsClassifier(n_neighbors=5)
knn_clf.fit(X_train, y_train)
y_pred_knn = knn_clf.predict(X_test)

# Evaluar KNN
print("KNN - Accuracy:", accuracy_score(y_test, y_pred_knn))

KNN - Accuracy: 0.72995


Comparar los resultados

In [50]:
print(f'Accuracy Logistic Regression: {accuracy_log}')
print(f'Accuracy Random Forest: {accuracy_rf}')
print(f'Accuracy XGBoost: {accuracy_xgb}')
print("Accuracy Gradient Boosting:", accuracy_score(y_test, y_pred_gb))
print("Accuracy KNN:", accuracy_score(y_test, y_pred_knn))

Accuracy Logistic Regression: 0.7095
Accuracy Random Forest: 0.76685
Accuracy XGBoost: 0.78515
Accuracy Gradient Boosting: 0.76995
Accuracy KNN: 0.72995


Despues de entrenar los cinco modelos: Regresión Logística, RandomForestClassifier, XGBClassifier, Gradient Boosting y K-Nearest Neighbors (KNN). Para seleccionar el mejor, usamos la métrica de accuracy.

En este caso el modelo de XGBoost ha tenido el mejor rendimiento en términos de accuracy (0.78515).

### Guardar el modelo XGBoost como archivo .pkl

In [51]:
joblib.dump(xgb_clf, '../Model/modelo_xgboost.pkl')

print("Modelo XGBoost guardado como 'modelo_xgboost.pkl'")

Modelo XGBoost guardado como 'modelo_xgboost.pkl'
